### ISSUE TITLE
gatsby-mdx 0.3.x upgrade unexpected token error

### ISSUE THREAD

**Issue by nwesthoff** [ChristopherBiscardi/gatsby-mdx#268]
**Describe the bug**
Any version prior to 0.3.0 works fine, any version after doesn't. I get an unexpected token error. All the code that is returned isn't mine (except for the graphql queries), which is why I'm not too sure where I have to start debugging.

**To Reproduce**
Run `gatsby develop`, get a bunch of errors:

```
success open and validate gatsby-configs — 0.011 s
success load plugins — 0.252 s
success onPreInit — 0.685 s
success delete html and css files from previous builds — 0.044 s
info One or more of your plugins have changed since the last time you ran Gatsby. As
a precaution, we're deleting your site's cache to ensure there's not any stale
data
success initialize cache — 0.017 s
success copy gatsby files — 0.051 s
success onPreBootstrap — 0.010 s
error Plugin gatsby-mdx returned an error


  SyntaxError: unknown: Unexpected token, expected "}" (30:16)
    28 |       <style>
    29 |         .embedVideo-iframe {
  > 30 |           border: 0
       |                 ^
    31 |         }
    32 |       </style><  m/MDXTag>
    33 |            </MDXTag>
  - index.js:3834 Object.raise
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3834:17

  - index.js:5142 Object.unexpected
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5142:16

  - index.js:5134 Object.expect
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5134:28

  - index.js:3406 Object.jsxParseExpressionContainer
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3406:10

  - index.js:3493 Object.jsxParseElementAt
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3493:34

  - index.js:3482 Object.jsxParseElementAt
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3482:32

  - index.js:3482 Object.jsxParseElementAt
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3482:32

  - index.js:3535 Object.jsxParseElement
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3535:17

  - index.js:3542 Object.parseExprAtom
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3542:19

  - index.js:5848 Object.parseExprSubscripts
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5848:23

  - index.js:5828 Object.parseMaybeUnary
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5828:21

  - index.js:5717 Object.parseExprOps
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5717:23

  - index.js:5690 Object.parseMaybeConditional
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5690:23

  - index.js:5635 Object.parseMaybeAssign
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5635:21

  - index.js:5587 Object.parseExpression
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5587:23

  - index.js:7559 Object.parseReturnStatement
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:7559:28

  - index.js:7237 Object.parseStatementContent
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:7237:21

  - index.js:7199 Object.parseStatement
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:7199:17

  - index.js:7757 Object.parseBlockOrModuleBlockBody
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:7757:25

  - index.js:7744 Object.parseBlockBody
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:7744:10

  - index.js:7733 Object.parseBlock
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:7733:10

  - index.js:6865 Object.parseFunctionBody
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:6865:24


error Plugin gatsby-mdx returned an error


  SyntaxError: unknown: Unexpected token (75:132)
    73 | <div class="gatsby-  highlight" data-language="tsx"><  9mpre class="language-tsx"><  39mcode class="language-tsx"><  span class="token keyword">  6mimport</span> <span  m class="token operator">*<  9m/span> <span class  ="token keyword">as</span  > React <span class="toke  n keyword">from</span> <  [39mspan class="token string">  'react'</span><span   class="token punctuation">;<  39m/span>
    74 |
  > 75 | <span class  m="token keyword">interface</span  39m> <span class="token class-  name">PrismTestProps</span  m> <span class="token punctuation"  [39m>{</span>
       |                                                        ^  39m
    76 |   name<span class="  token punctuation">:</span  m> <span class="token builtin"  >string</span><span  [39m class="token punctuation">;  3m</span>
    77 | <span class="token   punctuation">}</span>
    78 |

  - index.js:3834 Object.raise
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3834:17

  - index.js:5142 Object.unexpected
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5142:16

  - index.js:3327 Object.jsxParseIdentifier
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3327:12

  - index.js:3337 Object.jsxParseNamespacedName
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3337:23

  - index.js:3348 Object.jsxParseElementName
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3348:21

  - index.js:3433 Object.jsxParseOpeningElementAt
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3433:22

  - index.js:3466 Object.jsxParseElementAt
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3466:33

  - index.js:3535 Object.jsxParseElement
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3535:17

  - index.js:3542 Object.parseExprAtom
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3542:19

  - index.js:5848 Object.parseExprSubscripts
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5848:23

  - index.js:5828 Object.parseMaybeUnary
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5828:21

  - index.js:5717 Object.parseExprOps
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5717:23

  - index.js:5690 Object.parseMaybeConditional
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5690:23

  - index.js:5635 Object.parseMaybeAssign
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5635:21

  - index.js:5587 Object.parseExpression
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5587:23

  - index.js:3403 Object.jsxParseExpressionContainer
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3403:30

  - index.js:3493 Object.jsxParseElementAt
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3493:34

  - index.js:3482 Object.jsxParseElementAt
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3482:32

  - index.js:3482 Object.jsxParseElementAt
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3482:32

  - index.js:3482 Object.jsxParseElementAt
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3482:32

  - index.js:3482 Object.jsxParseElementAt
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3482:32

  - index.js:3535 Object.jsxParseElement
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3535:17


success source and transform nodes — 1.026 s
success building schema — 0.239 s
success createPages — 0.194 s
success createPagesStatefully — 0.035 s
success onPreExtractQueries — 0.003 s
success update schema — 0.125 s
success extract queries from components — 0.109 s
error
The GraphQL query from /Users/intern/Documents/git-repos/vanberlo-internship-report/src/components/listing.tsx failed.

Errors:
  unknown: Unexpected token (167:132)

    165 | <div class="gatsby-highlight" data-language="tsx"><pre class="language-tsx"><code class="language-tsx"><span class="token keyword">import</span> <span class="tokenoperator">*</span> <span class="token keyword">as</span> React <span class="token keyword">from</span> <span class="token string">'react'</span><span class="token punctuation">;</span>
    166 |
  > 167 | <span class="token keyword">interface</span> <span class="token class-name">PrismTestProps</span> <span class="token punctuation">{</span>
        |                                                       ^
    168 |   name<span class="token punctuation">:</span> <span class="token builtin">string</span><span class="token punctuation">;</span>
    169 | <span class="token punctuation">}</span>
    170 |

  GraphQL request (5:9)
  4:       node {
  5:         excerpt
             ^
  6:         frontmatter {
  ,unknown: Unexpected token, expected "}" (30:16)

    28 |       <style>
    29 |         .embedVideo-iframe {
  > 30 |           border: 0
       |                 ^
    31 |         }
    32 |       </style></MDXTag>
    33 |            </MDXTag>

  GraphQL request (5:9)
  4:       node {
  5:         excerpt
             ^
  6:         frontmatter {

Plugin:
  none
Query:
  query LISTING_QUERY {
    allMdx(limit: 10, sort: {fields: [frontmatter___date], order: ASC}) {
      edges {
        node {
          excerpt
          frontmatter {
            path
            title
            date(formatString: "MMMM DD, YYYY")
          }
        }
      }
    }
  }

success run graphql queries — 0.295 s — 17/17 57.97 queries/second
success write out page data — 0.003 s
success write out redirect data — 0.000 s
⢀ onPostBootstrapdone generating icons for manifest
success onPostBootstrap — 0.299 s

info bootstrap finished - 5.289 s

Warning: React version not specified in eslint-plugin-react settings. See https://github.com/yannickcr/eslint-plugin-react#configuration .
error UNHANDLED REJECTION


  SyntaxError: unknown: Unexpected token, expected "}" (32:16)
    30 |       <style>
    31 |         .embedVideo-iframe {
  > 32 |           border: 0
       |                 ^
    33 |         }
    34 |       </style><  m/MDXTag>
    35 |            </MDXTag>
  - index.js:3834 Object.raise
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3834:17

  - index.js:5142 Object.unexpected
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5142:16

  - index.js:5134 Object.expect
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5134:28

  - index.js:3406 Object.jsxParseExpressionContainer
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3406:10

  - index.js:3493 Object.jsxParseElementAt
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3493:34

  - index.js:3482 Object.jsxParseElementAt
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3482:32

  - index.js:3482 Object.jsxParseElementAt
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3482:32

  - index.js:3535 Object.jsxParseElement
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3535:17

  - index.js:3542 Object.parseExprAtom
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:3542:19

  - index.js:5848 Object.parseExprSubscripts
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5848:23

  - index.js:5828 Object.parseMaybeUnary
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5828:21

  - index.js:5717 Object.parseExprOps
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5717:23

  - index.js:5690 Object.parseMaybeConditional
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5690:23

  - index.js:5635 Object.parseMaybeAssign
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5635:21

  - index.js:5587 Object.parseExpression
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:5587:23

  - index.js:7559 Object.parseReturnStatement
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:7559:28

  - index.js:7237 Object.parseStatementContent
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:7237:21

  - index.js:7199 Object.parseStatement
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:7199:17

  - index.js:7757 Object.parseBlockOrModuleBlockBody
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:7757:25

  - index.js:7744 Object.parseBlockBody
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:7744:10

  - index.js:7733 Object.parseBlock
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:7733:10

  - index.js:6865 Object.parseFunctionBody
    [vanberlo-internship-report]/[@babel]/parser/lib/index.js:6865:24
```

**Expected behavior**
It to start the dev server without issues, like it would on 0.2.0.

**Additional context**
Running a 2015 MacBook Pro, MacOS Mojave 10.14.2.
npm is at 6.6.0
using [Gatsby Typescript MDX PrismJS starter](https://www.gatsbyjs.org/starters/tylergreulich/gatsby-typescript-mdx-prismjs-starter/)

**Comment by ar-stackrox** [comment]
This is happening to me as well.

**Comment by robertcoopercode** [comment]
Yes, when I upgraded to 0.3.x I get the same issue:

![image](https://user-images.githubusercontent.com/16786990/52185610-34eb1780-27ef-11e9-9497-ebe9673d150d.png)

For reference, I am also using `gatsby-remark-prismjs` which seems to be what is adding the markup that is causing the error.

**Comment by ar-stackrox** [comment]
Yeah, I had to downgrade for the time being :(

**Comment by robertcoopercode** [comment]
Ya, I'll stay at 0.2.x as well.

**Comment by Ehesp** [comment]
Also ran into this too. Got it working again using the suggested method above, however it's not a drop in (I think). 

`prism-react-renderer` requires the language to be passed in with a prop, and also things like line number highlighting, code titles no longer work as there's no handling/parsing of the class name. 

How come this broke from v0.2 -> v0.3?

**Comment by robertcoopercode** [comment]
@ChristopherBiscardi thanks for pointing out a solution to this issue, it was very helpful.

For those looking for a bit more detail on what is required to get `prism` working with `gatsby-mdx` here are the steps I took:

- Remove `gatsby-prism-remark` dependency and remove it from `gatsby-config.js`
- Install [`prism-react-renderer`](https://github.com/FormidableLabs/prism-react-renderer#prism)
- Pass a new `components` prop to the existing `<MDXProvider>` in the app with a custom `pre` element. You should also install `mdx-utils` to help compose the props that will get passed in the `pre` element.

Here's some code to help illustrate the above point:

```javascript
// Import helper function
import { preToCodeBlock } from 'mdx-utils';

// Import custom components
import Title from './Title';
import SubtitleH2 from './SubtitleH2';
import Paragraph from './Paragraph';
import Code from './Code';

// Defining the `MDXProvider` components prop value
const mdxComponents = {
    h1: props => <Title {...props} />,
    h2: props => <SubtitleH2 {...props} />,
    p: props => <Paragraph {...props} />,
    pre: props => {
        const preProps= preToCodeBlock(props);
        if (preProps) {
            return <Code {...preProps} />;
        } else {
            return <pre {...props} />;
        }
    },
};

// The rendering of the `MDXProvider` component and children
<MDXProvider components={mdxComponents}>
    <Main>{children}</Main>
</MDXProvider>;
```
- Define a custom component to be used for the `pre` elements that will be rendered using `prism-react-renderer`. Here's what my custom component looks like:

```javascript
import React from 'react';
import Highlight, { defaultProps } from 'prism-react-renderer';

const Code = ({ codeString, language, ...props }) => {
    return (
        <Highlight {...defaultProps} code={codeString} language={language} theme={undefined}>
            {({ className, tokens, getLineProps, getTokenProps }) => (
                <div className="gatsby-highlight">
                    <pre className={className}>
                        {tokens.map((line, i) => (
                            <div {...getLineProps({ line, key: i })}>
                                {line.map((token, key) => (
                                    <span {...getTokenProps({ token, key })} />
                                ))}
                            </div>
                        ))}
                    </pre>
                </div>
            )}
        </Highlight>
    );
};

export default Code;
```
Notice how i'm passing `undefined` the the `theme` prop in `Highlight`. This prevents the default styling to be applied to the rendered Prism code blocks and allows my own custom theme to get rendered.

**Comment by Ehesp** [comment]
@robertcoopercode thanks very much for that! The `mdx-utils` was my missing part, I was doing a regex to extract the language etc from the className 💩 

@ChristopherBiscardi no problem, appreciate the work. I'm building out the [react-native-firebase](https://github.com/invertase/react-native-firebase) docs with this and it's working very well :)

**Comment by ChristopherBiscardi** [comment]
This should be fixed in most cases as of `@mdx-js/mdx@0.20.1` and `gatsby-mdx@0.4.2`. Going to close this since it's fairly wide-ranging and if you still see this issue, please file a new issue with the specific instance of MDX content that is a problem.
